apps/nerves_hub_device/lib/nerves_hub_device/ssl.ex

```diff
@@ -1,111 +1,184 @@
+         true <- cn == device.identifier || :mismatched_cert,
+         {:ok, db_ca} <- check_known_ca(otp_cert),
+         {:ok, db_ca} <- check_expiration(db_ca),
+         true <- db_ca.org_id == device.org_id || :mismatched_org,
+         der = Certificate.to_der(otp_cert),
+         {:ok, _} <- :public_key.pkix_path_validation(db_ca.der, [der], []),
+         params = params_from_otp_cert(otp_cert) do
+      Devices.create_device_certificate(device, params)
+    end
+  end
+
+  defp check_common_name(otp_cert) do
+    case Certificate.get_common_name(otp_cert) do
+      cn when is_binary(cn) and byte_size(cn) > 0 ->
+        {:ok, cn}
 
       _ ->
-        {:fail, :unknown_ca}
+        :missing_common_name
     end
   end
 
-  def verify_device(certificate) do
-    case Devices.get_device_certificate_by_x509(certificate) do
-      {:ok, cert} ->
-        # TODO: Remove once enough time has allowed existing DERs to be captured ¬
-        Devices.update_device_certificate(cert, %{
-          last_used: DateTime.utc_now(),
-          der: Certificate.to_der(certificate)
-        })
+  defp check_known_ca(otp_cert) do
+    Certificate.get_aki(otp_cert)
+    |> Devices.get_ca_certificate_by_ski()
+    |> case do
+      {:ok, db_ca} ->
+        # Mark that this CA cert was used
+        Devices.update_ca_certificate(db_ca, %{last_used: DateTime.utc_now()})
 
       _ ->
-        with aki <- Certificate.get_aki(certificate),
-             {:ok, %CACertificate{org_id: org_id}} <- Devices.get_ca_certificate_by_ski(aki),
-             {:ok, org} <- Accounts.get_org(org_id),
-             identifier <- Certificate.get_common_name(certificate),
-             {:ok, device} <- Devices.get_device_by_identifier(org, identifier) do
-          attempt_registration(device, certificate)
-        else
-          _e -> :error
-        end
+        :unknown_ca
     end
   end
 
-  defp attempt_registration(device, certificate) do
-    with [] <- Devices.get_device_certificates(device),
-         serial <- Certificate.get_serial_number(certificate),
-         aki <- Certificate.get_aki(certificate),
-         ski <- Certificate.get_ski(certificate),
-         der <- Certificate.to_der(certificate),
-         {not_before, not_after} <- Certificate.get_validity(certificate),
-         params <- %{
-           serial: serial,
-           aki: aki,
-           ski: ski,
-           not_before: not_before,
-           not_after: not_after,
-           der: der
-         },
-         {:ok, db_cert} <- Devices.create_device_certificate(device, params) do
-      {:ok, db_cert}
+  defp check_expiration(db_cert) do
+    now = DateTime.utc_now()
+    is_after? = DateTime.compare(now, db_cert.not_after) == :gt
+    is_before? = DateTime.compare(now, db_cert.not_before) != :gt
+
+    if is_before? or is_after? do
+      # Maybe should be :cert_expired ?
+      :invalid_issuer
     else
-      _e -> :error
+      {:ok, db_cert}
```

Comment:
```suggestion
      :ok
```
This changes some other code above. It looked like `db_cert` was going to be rebound above. To avoid any confusion, I figured it would be better to just return `:ok` here.